sql: add LTREE datum with valueside encoding

paulniziolek opened the pull request:
#### sql: add LTREE datum with basic functionality and pgwire encoding

The ltree type LTREE is recognized as valid types by the parser, but is still
mostly featureless. LTREE is supported as a table column and value, as they are
parsed into *DLTree datums. An LTREE row is represented by a dot-separated Path
of string labels, with some constraints on length and characters accepted. For
example, `'Top.Middle.Bottom'::LTREE` and `'foo-bar.baz_bop'` are valid LTREE
Paths.

Informs: cockroachdb#44657
Epic: CRDB-148

Release note: None

#### sql: add LTREE valueside encoding LTREE datums are valueside encoded by their byte representation including path separators.

Informs: cockroachdb#44657
Epic: CRDB-148

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/151439)
<!-- Reviewable:end -->

paulniziolek commented:
Note to reviewers: This is stacked on https://github.com/cockroachdb/cockroach/pull/149912

paulniziolek committed:
sql: add LTREE datum with basic functionality and pgwire encoding

The ltree type LTREE is recognized as valid types by the parser, but is still
mostly featureless. LTREE is supported as a table column and value, as they are
parsed into *DLTree datums. An LTREE row is represented by a dot-separated Path
of string labels, with some constraints on length and characters accepted. For
example, `'Top.Middle.Bottom'::LTREE` and `'foo-bar.baz_bop'` are valid LTREE
Paths.

Informs: cockroachdb#44657
Epic: CRDB-148

Release note: None

paulniziolek commented on pkg/util/encoding/encoding.go:
```diff
@@ -136,10 +136,18 @@ const (
 	jsonKeyTerminator           byte = 0x00
 	jsonKeyDescendingTerminator byte = 0xFF
 
+	// Terminators for LTREE Key encoding.
+	ltreeKeyTerminator           byte = 0x0F
+	ltreeKeyDescendingTerminator byte = 0xF1
```
these were chosen somewhat arbitrarily. the byte space reserved for ltree labels is from 45-122. i'm not sure if LTREEs w/ 0x00 and 0xFF would collide with other encodings such as ltree within arrays, but I don't think so.

paulniziolek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151439)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/randgen/types_test.go` line 40 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4AHf11bpjDW1H1o8c:-OX4AHf11bpjDW1H1o8d:b-sg8woq) ([raw file](https://github.com/cockroachdb/cockroach/blob/97c728cea281f469a3a00686fbb36ccdc780edd4/pkg/sql/randgen/types_test.go#L40)):*
> ```Go
> 			continue loop
> 		case types.LQueryFamily, types.LTXTQueryFamily:
> 			continue loop
> ```

nit: add comment
___
*[`pkg/util/encoding/encoding.go` line 1882 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX49NrY1GFM8vr9tt59:-OX49NrY1GFM8vr9tt5A:b-3kdc74) ([raw file](https://github.com/cockroachdb/cockroach/blob/97c728cea281f469a3a00686fbb36ccdc780edd4/pkg/util/encoding/encoding.go#L1882)):*
> ```Go
> 	PGVector           Type = 44
> 	LTree              Type = 45
> 	LTreeDesc          Type = 46
> ```

I could potentially combine these into just `LTree`, and have `slowPeekType` below return `LTree` for both asc and desc markers. 
___
*[`pkg/sql/pgwire/types.go` line 893 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX48Sw00baoG1R-YzEh:-OX48Sw00baoG1R-YzEi:bomvf0n) ([raw file](https://github.com/cockroachdb/cockroach/blob/97c728cea281f469a3a00686fbb36ccdc780edd4/pkg/sql/pgwire/types.go#L893)):*
> ```Go
> 		b.writeByte(1)
> 
> 		v.LTree.ForEachLabel(func(i int, label string) {
> ```

nit: use `v.LTree.FormatToBuffer` instead


<!-- Sent from Reviewable.io -->

mgartner reviewed (changes requested):
Flushing my comments for the first 2 commits. You might want to put the third in a separate PR which you iterate on the first two.

> Reviewed 63 of 67 files at r1, 1 of 1 files at r2, 13 of 13 files at r5, 6 of 6 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-:-OX4Zf2V5Dk5qBeRZwXv:bv9a2ei)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/util/encoding/encoding.go` line 2847 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4TJitEBvMYHvxG6vf:-OX4TJitEBvMYHvxG6vg:buywmth) ([raw file](https://github.com/cockroachdb/cockroach/blob/12321f2745d6e6b68533c4d13ff6936a303096e6/pkg/util/encoding/encoding.go#L2847)):*
> ```Go
> // value tag but with a length prefix, appends it to the supplied buffer, and
> // returns the final buffer.
> func EncodeLTreeValue(appendTo []byte, colIDDelta uint32, data []byte) []byte {
> ```

nit: I think these encoding changes are in the wrong commit?
___
*[`pkg/sql/parser/testdata/select_exprs` line 2159 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4SMYr85jcQvUFU2nH:-OX4SMYr85jcQvUFU2nI:bwfxkc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/12321f2745d6e6b68533c4d13ff6936a303096e6/pkg/sql/parser/testdata/select_exprs#L2159)):*
> ```
> 
> parse
> SELECT '$'::LTREE
> ```

nit: these `'$'` aren't valid LTREEs. It doesn't really matter since we're only parsing the SQL here, not the LTREE, but I think it's nice for future folks who stumble upon these tests to see values that actually make sense
___
*[`pkg/util/encoding/encoding.go` line 2855 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4VKTSDlzLV4jjhG0D:-OX4VKTSDlzLV4jjhG0E:bm35uxu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9a090d880df1246ede03c57da0a30c6a845abd/pkg/util/encoding/encoding.go#L2855)):*
> ```Go
> // buffer, and returns the final buffer.
> func EncodeUntaggedLTreeValue(appendTo []byte, l ltree.T) []byte {
> 	appendTo = EncodeUntaggedBytesValue(appendTo, UnsafeConvertStringToBytes(l.String()))
> ```

We should avoid the allocation for `String` here. Can we use the Format method or something similar to append bytes directly to appendTo? We do that here: https://github.com/cockroachdb/cockroach/blob/19afc8b16c6e13cdb249e8a0c8db1c68bab568ca/pkg/util/encoding/encoding.go#L2808
___
*[`pkg/util/encoding/encoding.go` line 3221 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4XaN5-CE5HsKVVjOM:-OX4XaN5-CE5HsKVVjON:byr5pov) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9a090d880df1246ede03c57da0a30c6a845abd/pkg/util/encoding/encoding.go#L3221)):*
> ```Go
> 		return b, l, err
> 	}
> 	l, err = ltree.ParseLTree(UnsafeConvertBytesToString(data))
> ```

I don't think it is safe to pass an unsafe-bytes-converted-to-string here. `ParseLTree` sets its internal labels to the result of `strings.Split`. `strings.Split` returns a `[]string` where each string is a slice of the input string, not a newly allocated string. So the resulting `ltree.T` will have a `[]string` where the strings are slices of the `data` byte slice, which is mutable and could be written over. 

It looks like DecodeUntaggedBytesValue also reuses the original slice: https://github.com/cockroachdb/cockroach/blob/19afc8b16c6e13cdb249e8a0c8db1c68bab568ca/pkg/util/encoding/encoding.go#L2973

But the users of that function convert the []bytes back to a string which copies the bytes to a new location in memory, ensuring that the string remains immutable, e.g.: https://github.com/cockroachdb/cockroach/blob/d847f6e6abc21f237c59297ac4d5a654a346e46c/pkg/sql/rowenc/valueside/decode.go#L77

So I think we should do this instead: `ltree.ParseLTree(string(data))`.
___
*[`pkg/sql/pgwire/types.go` line 883 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4OBts-_DM9G0rRgpJ:-OX4OBttA3uvfv2TU0u4:bds2wa3) ([raw file](https://github.com/cockroachdb/cockroach/blob/12321f2745d6e6b68533c4d13ff6936a303096e6/pkg/sql/pgwire/types.go#L883)):*
> ```Go
> 		// The first byte is always used for the version number
> 		size := int32(1)
> 		v.LTree.ForEachLabel(func(i int, label string) {
> ```

nit: can you use `v.LTree.ByteSize()` here instead?
___
*[`pkg/sql/pgwire/types.go` line 893 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4Ofyy40MLDEi8aPU_:-OX4Ofyy40MLDEi8aPUa:b-w8fuqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/12321f2745d6e6b68533c4d13ff6936a303096e6/pkg/sql/pgwire/types.go#L893)):*
> ```Go
> 		b.writeByte(1)
> 
> 		v.LTree.ForEachLabel(func(i int, label string) {
> ```

nit: can you reuse the format method here, like: https://github.com/cockroachdb/cockroach/blob/2dda5fd4f33f3b7adac591e406e16d41fd350208/pkg/sql/pgwire/types.go#L203-L205

and then remove `ForEachLabel` unless you need it elsewhere
___
*[`pkg/sql/sem/tree/eval.go` line 1916 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4Ql8d5bFdwU5fAs8w:-OX4Ql8d5bFdwU5fAs8x:bn7ylcd) ([raw file](https://github.com/cockroachdb/cockroach/blob/12321f2745d6e6b68533c4d13ff6936a303096e6/pkg/sql/sem/tree/eval.go#L1916)):*
> ```Go
> 		makeEvalTupleIn(types.Interval, volatility.Leakproof),
> 		makeEvalTupleIn(types.Jsonb, volatility.Leakproof),
> 		makeEvalTupleIn(types.LTree, volatility.Leakproof),
> ```

It's odd to mark this leakproof when the other operators are immutable. All leakproof ops are marked immutable and not leakproof in postres, so let's go with immutable here.

```
marcus=# select proname, provolatile, proleakproof from pg_proc where proname like '%ltree%';
         proname          | provolatile | proleakproof
--------------------------+-------------+--------------
 subltree                 | i           | f
 ltree_in                 | i           | f
 ltree_out                | i           | f
 ltree_cmp                | i           | f
 ltree_lt                 | i           | f
 ltree_le                 | i           | f
 ltree_eq                 | i           | f
 ltree_ge                 | i           | f
 ltree_gt                 | i           | f
 ltree_ne                 | i           | f
 ltree2text               | i           | f
 text2ltree               | i           | f
 ltree_isparent           | i           | f
 ltree_risparent          | i           | f
 _ltree_risparent         | i           | f
 ltree_addltree           | i           | f
 ltree_addtext            | i           | f
 ltree_textadd            | i           | f
 ltreeparentsel           | i           | f
 ltree_gist_in            | i           | f
 ltree_gist_out           | i           | f
 ltree_consistent         | i           | f
 ltree_compress           | i           | f
 ltree_decompress         | i           | f
 ltree_penalty            | i           | f
 ltree_picksplit          | i           | f
 ltree_union              | i           | f
 ltree_same               | i           | f
 _ltree_isparent          | i           | f
 _ltree_r_isparent        | i           | f
 _ltree_r_risparent       | i           | f
 _ltree_extract_isparent  | i           | f
 _ltree_extract_risparent | i           | f
 _ltree_consistent        | i           | f
 _ltree_compress          | i           | f
 _ltree_penalty           | i           | f
 _ltree_picksplit         | i           | f
 _ltree_union             | i           | f
 _ltree_same              | i           | f
 ltree_recv               | i           | f
 ltree_send               | i           | f
 ltree_gist_options       | i           | f
 _ltree_gist_options      | i           | f
(43 rows)
```
___
*[`pkg/sql/logictest/testdata/logic_test/ltree` line 2 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4RT198fCwQvplvNru:-OX4RT198fCwQvplvNrv:b-30l9zi) ([raw file](https://github.com/cockroachdb/cockroach/blob/12321f2745d6e6b68533c4d13ff6936a303096e6/pkg/sql/logictest/testdata/logic_test/ltree#L2)):*
> ```
> statement ok
> CREATE TABLE l (lt LTREE);
> ```

nit: I think these tests are in the wrong commit - you need valueside encoding for them to work, right?
___
*[`pkg/sql/logictest/testdata/logic_test/ltree` line 84 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4Rl0O2wh6J95pTf8_:-OX4Rl0O2wh6J95pTf8a:b-cc8cz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/12321f2745d6e6b68533c4d13ff6936a303096e6/pkg/sql/logictest/testdata/logic_test/ltree#L84)):*
> ```
> # In the future, LQUERY should be supported as a column type and usable type.
> query error cannot use lquery type as table column
> CREATE TABLE lq (lq LQUERY);
> ```

I see, the LQUERY types are needed in order to have an error message like this. It's fine to keep, if you'd like. I think I'd prefer to leave it all out so we don't have mostly-dead code lying around. I don't think this error message is all that much more helpful either—it's potentially confusing because it sounds like LQUERY is supported, just not as a table column. Up to you though.
___
*[`pkg/sql/sem/tree/datum.go` line 4551 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4PSPg9-iE_qCNTPNb:-OX4PSPg9-iE_qCNTPNc:b-db5662) ([raw file](https://github.com/cockroachdb/cockroach/blob/12321f2745d6e6b68533c4d13ff6936a303096e6/pkg/sql/sem/tree/datum.go#L4551)):*
> ```Go
> func (d *DLTree) Prev(ctx context.Context, cmpCtx CompareContext) (Datum, bool) {
> 	prevLTree, ok := d.LTree.Prev()
> 	return NewDLTree(prevLTree), ok
> ```

nit: we don't want to call `NewDLTree` if `ok=false`, so we should have a conditional:

```go
if !ok {
    return nil, false
}
``` 
___
*[`pkg/sql/sem/tree/datum.go` line 4571 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4Povp5wBzj-XA8t3Q:-OX4Povp5wBzj-XA8t3R:bh6ogdb) ([raw file](https://github.com/cockroachdb/cockroach/blob/12321f2745d6e6b68533c4d13ff6936a303096e6/pkg/sql/sem/tree/datum.go#L4571)):*
> ```Go
> // Min implements the Datum interface.
> func (d *DLTree) Min(ctx context.Context, cmpCtx CompareContext) (Datum, bool) {
> 	return NewDLTree(ltree.Empty), false
> ```

nit: return `true` here, right?
___
*[`pkg/sql/sem/tree/datum.go` line 6357 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4Q5RDA4dPH-XvbM57:-OX4Q5RDA4dPH-XvbM58:b-dwa1d7) ([raw file](https://github.com/cockroachdb/cockroach/blob/12321f2745d6e6b68533c4d13ff6936a303096e6/pkg/sql/sem/tree/datum.go#L6357)):*
> ```Go
> 	types.EnumFamily:           {unsafe.Sizeof(DEnum{}), variableSize},
> 	types.LTreeFamily:          {unsafe.Sizeof(DLTree{}), variableSize},
> 	// TODO(paulniziolek): This is suspicious.
> ```

nit: clarify this comment or remove it
___
*[`pkg/sql/types/types.go` line 580 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4R4dlDpr8FOrKQW-b:-OX4R4dlDpr8FOrKQW-c:b-pr4wa3) ([raw file](https://github.com/cockroachdb/cockroach/blob/12321f2745d6e6b68533c4d13ff6936a303096e6/pkg/sql/types/types.go#L580)):*
> ```Go
> 
> 	// LQuery is the type for a variable representing a query in a label tree.
> 	LQuery = &T{
> ```

nit: I would remove LQuery and LTXTQuery if they aren't being used.
___
*[`pkg/sql/logictest/testdata/logic_test/pg_catalog` line 3805 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4SGlE7iVaXROj4ok7:-OX4SGlE7iVaXROj4ok8:b-gjgjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2ba9a9f285c874f7cf9e8c2d449e924eb9f6b40/pkg/sql/logictest/testdata/logic_test/pg_catalog#L3805)):*
> ```
> 
> subtest pg_catalog.pg_prepare_statement
> ```

Looks like you accidentally deleted a bunch of tests here.


<!-- Sent from Reviewable.io -->

paulniziolek committed:
sql: add LTREE valueside encoding
LTREE datums are valueside encoded by their byte representation including path
separators.

Informs: cockroachdb#44657
Epic: CRDB-148

Release note: None

paulniziolek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151439)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/pgwire/types.go` line 893 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4Ofyy40MLDEi8aPU_:-OX5fS4m7mINGGgupVAd:b5hhoy4) ([raw file](https://github.com/cockroachdb/cockroach/blob/12321f2745d6e6b68533c4d13ff6936a303096e6/pkg/sql/pgwire/types.go#L893)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: can you reuse the format method here, like: https://github.com/cockroachdb/cockroach/blob/2dda5fd4f33f3b7adac591e406e16d41fd350208/pkg/sql/pgwire/types.go#L203-L205

and then remove `ForEachLabel` unless you need it elsewhere
</blockquote></details>

I will hold onto the `ForEachLabel` implementation here as it will be used by LTREE ops/funcs. 
___
*[`pkg/sql/sem/tree/eval.go` line 1916 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4Ql8d5bFdwU5fAs8w:-OX5ff7D2nOqydHMHz9X:b8hqpzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/12321f2745d6e6b68533c4d13ff6936a303096e6/pkg/sql/sem/tree/eval.go#L1916)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It's odd to mark this leakproof when the other operators are immutable. All leakproof ops are marked immutable and not leakproof in postres, so let's go with immutable here.

```
marcus=# select proname, provolatile, proleakproof from pg_proc where proname like '%ltree%';
         proname          | provolatile | proleakproof
--------------------------+-------------+--------------
 subltree                 | i           | f
 ltree_in                 | i           | f
 ltree_out                | i           | f
 ltree_cmp                | i           | f
 ltree_lt                 | i           | f
 ltree_le                 | i           | f
 ltree_eq                 | i           | f
 ltree_ge                 | i           | f
 ltree_gt                 | i           | f
 ltree_ne                 | i           | f
 ltree2text               | i           | f
 text2ltree               | i           | f
 ltree_isparent           | i           | f
 ltree_risparent          | i           | f
 _ltree_risparent         | i           | f
 ltree_addltree           | i           | f
 ltree_addtext            | i           | f
 ltree_textadd            | i           | f
 ltreeparentsel           | i           | f
 ltree_gist_in            | i           | f
 ltree_gist_out           | i           | f
 ltree_consistent         | i           | f
 ltree_compress           | i           | f
 ltree_decompress         | i           | f
 ltree_penalty            | i           | f
 ltree_picksplit          | i           | f
 ltree_union              | i           | f
 ltree_same               | i           | f
 _ltree_isparent          | i           | f
 _ltree_r_isparent        | i           | f
 _ltree_r_risparent       | i           | f
 _ltree_extract_isparent  | i           | f
 _ltree_extract_risparent | i           | f
 _ltree_consistent        | i           | f
 _ltree_compress          | i           | f
 _ltree_penalty           | i           | f
 _ltree_picksplit         | i           | f
 _ltree_union             | i           | f
 _ltree_same              | i           | f
 ltree_recv               | i           | f
 ltree_send               | i           | f
 ltree_gist_options       | i           | f
 _ltree_gist_options      | i           | f
(43 rows)
```
</blockquote></details>

Yeah.. this one I didn't exactly check. In a land full of Leakproofs, I misjudged and thought this must be Leakproof too. Thanks for showing me this query though! 
___
*[`pkg/util/encoding/encoding.go` line 2855 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4VKTSDlzLV4jjhG0D:-OX5gDQw6Ez0BajMqDlO:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9a090d880df1246ede03c57da0a30c6a845abd/pkg/util/encoding/encoding.go#L2855)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

We should avoid the allocation for `String` here. Can we use the Format method or something similar to append bytes directly to appendTo? We do that here: https://github.com/cockroachdb/cockroach/blob/19afc8b16c6e13cdb249e8a0c8db1c68bab568ca/pkg/util/encoding/encoding.go#L2808
</blockquote></details>

Done. 
___
*[`pkg/util/encoding/encoding.go` line 3221 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4XaN5-CE5HsKVVjOM:-OX5gJMO7gSj2qR8iu84:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9a090d880df1246ede03c57da0a30c6a845abd/pkg/util/encoding/encoding.go#L3221)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I don't think it is safe to pass an unsafe-bytes-converted-to-string here. `ParseLTree` sets its internal labels to the result of `strings.Split`. `strings.Split` returns a `[]string` where each string is a slice of the input string, not a newly allocated string. So the resulting `ltree.T` will have a `[]string` where the strings are slices of the `data` byte slice, which is mutable and could be written over. 

It looks like DecodeUntaggedBytesValue also reuses the original slice: https://github.com/cockroachdb/cockroach/blob/19afc8b16c6e13cdb249e8a0c8db1c68bab568ca/pkg/util/encoding/encoding.go#L2973

But the users of that function convert the []bytes back to a string which copies the bytes to a new location in memory, ensuring that the string remains immutable, e.g.: https://github.com/cockroachdb/cockroach/blob/d847f6e6abc21f237c59297ac4d5a654a346e46c/pkg/sql/rowenc/valueside/decode.go#L77

So I think we should do this instead: `ltree.ParseLTree(string(data))`.
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/ltree` line 84 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4Rl0O2wh6J95pTf8_:-OX5gNv02bP7CGDHmLkw:b-hpyub3) ([raw file](https://github.com/cockroachdb/cockroach/blob/12321f2745d6e6b68533c4d13ff6936a303096e6/pkg/sql/logictest/testdata/logic_test/ltree#L84)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I see, the LQUERY types are needed in order to have an error message like this. It's fine to keep, if you'd like. I think I'd prefer to leave it all out so we don't have mostly-dead code lying around. I don't think this error message is all that much more helpful either—it's potentially confusing because it sounds like LQUERY is supported, just not as a table column. Up to you though.
</blockquote></details>

Done. I'll remove it because I am also not a fan of dead code & I realize that these querying types might not need to be supported anytime soon. 
___
*[`pkg/sql/logictest/testdata/logic_test/pg_catalog` line 3805 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX4SGlE7iVaXROj4ok7:-OX5gZGS8BwsViSUDm42:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2ba9a9f285c874f7cf9e8c2d449e924eb9f6b40/pkg/sql/logictest/testdata/logic_test/pg_catalog#L3805)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Looks like you accidentally deleted a bunch of tests here.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!

> Reviewed 33 of 33 files at r8, 13 of 13 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-:-OX939gS1i28F_5xhnP2:bcmw7ss)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @paulniziolek)

___
*[`pkg/util/encoding/encoding.go` line 2857 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX91V7V8orW3LbcohbB:-OX91V7V8orW3LbcohbC:bntdfll) ([raw file](https://github.com/cockroachdb/cockroach/blob/9741c6b59f2de27cea18e18672c93f95ef2096d8/pkg/util/encoding/encoding.go#L2857)):*
> ```Go
> 	var buf bytes.Buffer
> 	l.FormatToBuffer(&buf)
> 	appendTo = EncodeUntaggedBytesValue(appendTo, buf.Bytes())
> ```

Nice. The string allocation is gone, but there is still at least 1 allocation for the buffer. I was thinking of something like this to get rid of the buffer allocation(s):

```go
	appendTo = EncodeNonsortingUvarint(appendTo, uint64(l.ByteSize()))
	l.ForEachLabel(func(i int, path string) {
		if i > 0 {
			appendTo = append(appendTo, ltree.PathSeparatorByte)
		}
		appendTo = append(appendTo, path...)
	})
	return appendTo
```

The resulting encoding is the same, so we can still use `DecodeUntaggedBytesValue` in `DecodeUntaggedLTreeValue`. Might be good to leave a commit that this logic matches the encoding of `EncodeUntaggedBytesValue`—a variable-encoded length followed by the content bytes.


<!-- Sent from Reviewable.io -->

mgartner commented:
Don't forget to update the PR description—it still includes the third commit.

paulniziolek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151439)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/util/encoding/encoding.go` line 2857 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX91V7V8orW3LbcohbB:-OX971F57Ub2ctKNPHwT:b-wy01ny) ([raw file](https://github.com/cockroachdb/cockroach/blob/9741c6b59f2de27cea18e18672c93f95ef2096d8/pkg/util/encoding/encoding.go#L2857)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Nice. The string allocation is gone, but there is still at least 1 allocation for the buffer. I was thinking of something like this to get rid of the buffer allocation(s):

```go
	appendTo = EncodeNonsortingUvarint(appendTo, uint64(l.ByteSize()))
	l.ForEachLabel(func(i int, path string) {
		if i > 0 {
			appendTo = append(appendTo, ltree.PathSeparatorByte)
		}
		appendTo = append(appendTo, path...)
	})
	return appendTo
```

The resulting encoding is the same, so we can still use `DecodeUntaggedBytesValue` in `DecodeUntaggedLTreeValue`. Might be good to leave a commit that this logic matches the encoding of `EncodeUntaggedBytesValue`—a variable-encoded length followed by the content bytes.
</blockquote></details>

I think I would have to write something like this below, and I think the cast from string to []byte causes an allocation anyway? Or I would have to be more verbose when adding the `PathSeparator` string via another loop. For now, I'll merge this, if this needs to be fixed I can follow up on this in the upcoming PRs. 

But it also seems that in Go 1.22, type casting from string to []byte has been optimized in some way regarding allocations. 



_Code snippet:_
```Go
appendTo = append(appendTo, []byte(ltree.PathSeparator)...)
```


<!-- Sent from Reviewable.io -->

paulniziolek commented:
bors r+

mgartner commented:
*[`pkg/util/encoding/encoding.go` line 2857 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/151439#-OX91V7V8orW3LbcohbB:-OX9CdQGCMjv93WIM_Er:bxyhz0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/9741c6b59f2de27cea18e18672c93f95ef2096d8/pkg/util/encoding/encoding.go#L2857)):*
> For now, I'll merge this, if this needs to be fixed I can follow up on this in the upcoming PRs.

SGTM

I think having two constants exported in `ltree` would work well:

```go
const (
	PathSeparator       = '.'
	PathSeparatorString = string(PathSeparator...)
)
```



<!-- Sent from Reviewable.io -->

mgartner commented:
Try rebasing on latest master? Assuming your upstream remote is called `upstream`:

```
git fetch upstream
git rebase upstream/master
./dev gen testlogic
```

yuzefovich commented:
For awareness, missing key encoding support results in several failures, including in tests owned by other teams: [search](https://github.com/cockroachdb/cockroach/issues?q=sort%3Aupdated-desc%20is%3Aissue%20is%3Aopen%20%22unable%20to%20encode%20table%20key%3A%20*tree.DLTree%22).

I'll see whether we can silence the error by marking it a non-assertion failure for LTREE.

mgartner commented:
Thanks for the heads-up. We have other types that cannot be key-encoded, so it seems like we just missed a spot or two to denote that.
